qutebrowser/browser/webelem.py

```diff
@@ -48,16 +48,30 @@ class OrphanedError(Error):
     """Raised when a webelement's parent has vanished."""
 
 
-def css_selector(group: str, url: QUrl) -> str:
-    """Get a CSS selector for the given group/URL."""
+SPECIAL_SELECTOR_MAP = {
+    "__qute_scrollable": "scrollable"
+}
+
+
+def css_selector(group: str, url: QUrl) -> typing.Tuple[str, typing.List[str]]:
+    """Get CSS and special selectors for the given group/URL."""
     selectors = config.instance.get('hints.selectors', url)
     if group not in selectors:
         selectors = config.val.hints.selectors
 
         if group not in selectors:
             raise Error("Undefined hinting group {!r}".format(group))
 
-    return ','.join(selectors[group])
+    selectors = selectors[group]
+    css_selectors = []  # type: typing.List[str]
+    special_selectors = []  # type: typing.List[str]
+    for s in selectors:
```

Comment:
You could use a `set()` to simplify a bit

```py
selectors = set(selectors[group])
css_selectors = selectors.difference(SPECIAL_SELECTOR_MAP)
special_selectors = selectors.intersection(SPECIAL_SELECTOR_MAP)
```

Reply:
PS This still iterates over `selectors[group]` during the creation of the set. But I think a set is a much more appropriate representation of selectors than a list anyway, as their order doesn't matter. So if the config system stored selectors as sets, the set would not have to be crated here, and you'd only have to loop over `SPECIAL_SELECTOR_MAP`. Don't know if it is possible with the current config system, though.

Reply:
PPS You could change the order and save a few iterations when `selectors[group]` does not contain any of the special selectors.

```py
special_selectors = selectors.intersection(SPECIAL_SELECTORS_MAP)
css_selectors = selectors.difference(special_selectors)
```

Reply:
It would be very nice to have this as a set already. However, I don't think we can make that change very easily at the moment because it would break backwards compatibility. I like using set operations though I'll change it :]

Reply:
What needs to be backward compatible? Only thing I can think of is user config, and there you could still accept a list (Are sets even a thing in YAML?) and then convert it to a set internally.

Reply:
Yah, I was referring to the user configuration. We could definitely do something like that (expose a list on the user config and a set internally), but right now the user config and internal code follows the same paths. Maybe this would be a nice feature to have longer term, but definitely not worth it at the minute. 

I also took a quick look at the time spent here, and it seems like this is actually way slower (I'm guessing due to the conversion and the double hash-table lookup), so I'll undo the last commit, even though it's way more readable and nice :/. Maybe if this is eventually stored as a set we can change this back.

Reply:
Oh in that case it's fair to just leave it be for now.

Yeah the slowdown was to be expected. You could argue, though, that it doesn't matter, since this code is executed once when the user initiates hinting (or am I wrong here?), and I can't imagine there would be a perceptible difference between a single run of the loop vs the set. Do whatever you see fit, though. It was a fun exercise to come up with a readable version of the loop :)

Reply:
PS I see you did a force push of the previous commit to undo the commit. Did you know about `git revert`? It's super useful because it makes a new commit that undoes the reverted commit, where you can explain in the commit message why this behaviour was reverted. And as a bonus you keep the original commit in the history, so you can "easily" find it in case you want to consider it again in the future.

Reply:
> We could definitely do something like that (expose a list on the user config and a set internally), but right now the user config and internal code follows the same paths. Maybe this would be a nice feature to have longer term, but definitely not worth it at the minute.

Should be easy to add a new configtype which takes a list from YAML/config.py but returns a set to Python. See e.g. `QtColor` which takes a string but returns a `QColor` object to Python.

> I also took a quick look at the time spent here, and it seems like this is actually way slower (I'm guessing due to the conversion and the double hash-table lookup), so I'll undo the last commit, even though it's way more readable and nice :/

What does "way slower" mean in numbers? Like @olmokramer says, I doubt it makes a noticable difference when hinting, or does it?

Reply:
For me at least, it was about an order of magnitude faster ~(50us vs 5 us) with the default hint targets. This code is a regression (that's why I originally pointed this out), and I really don't like adding regressions. Even 5us for something this simple is a huge regression, imo. 

Does it make a difference in this case? Probably not, but that's only because we're busy making copies of dictionaries (instead of sets) in the config system and updating the binding dict instead. Hinting is still much slower than in other plugins (just give vimium a try on a site with many hits).

If we are going to make a special config type, I think we can probably do all this processing ahead of time and just return the tuple when requested. There's no reason for this loop or string join to run every time we hint.

I know about git revert, I just don't like cluttering the history with WIP commits. This PRs history is a bit convoluted, but I can at least try to keep the new stuff neat :)